EIP-5345  Walletconnect Silent-Signing Extension (#5345)
ethereum/EIPs: EIPS/eip-5345.md

---
eip: 5345
title: Walletconnect Silent-Signing Extension
description: Capability of transaction signing without user interaction for a period of time for walletconnect
author: Stanley Wu (@fruit37), Mücahit Büyükyılmaz (@anndro), Muhammed Emin Aydın (@muhammedea)
discussions-to: https://ethereum-magicians.org/t/walletconnect-silent-signing-extension/10137
status: Draft
type: Standards Track
category: Interface
created: 2022-07-26
---

## Abstract

> For mobile applications that do a lot of transactions, switching between wallet and application is a bad user experience, so in this proposal, we propose a way of signing and sending transactions by wallet without r intervention. This should be done for a period of time with user consent. We call this feature, Silent Signing.

Reviewer: ```suggestion
For mobile applications that do a lot of transactions, switching between wallet and application is a bad user experience, so in this proposal, we propose a way of signing and sending transactions by wallet without intervention. This should be done for a period of time with user consent. We call this feature, Silent Signing.
```

Author: thanks

WalletConnect is a protocol that allows web3 apps to connect blockchain wallets. The protocol is built on top of a WebSocket connection, between app-bridge and bridge-wallet. The protocol establishes a remote connection between two apps and/or devices. These payloads are symmetrically encrypted through a shared key between the two peers. Payload exchange between app and wallet goes over through Bridge server. As the payload is encrypted, the Bridge server redirects packages between app-wallet by topic id, without checking package details.

Decrypted package contains standard RPC commands and parameters. WalletConnect client support set of RPC commands and only that commands are usable with official clients. Please note that these RPC commands are different from Ethereum RPC commands. These commands only define wallet actions.

***Supported WalletConnect client commands***

> * `personal_sign`

Reviewer 2: ```suggestion

* `personal_sign`
```

* `eth_sign`
* `eth_signTypedData`
* `eth_sendTransaction`
* `eth_signTransaction`
* `eth_sendRawTransaction`

As a nature of the Bridge server, it redirects payload packages directly without knowing which RPC call has been made. For that reason, we don't need to change anything on the Bridge server, but WalletConnect clients.

### Silent Signing User Flow

Silent Signing process is like the following:

* First, the application sends a silent signing request to the wallet. The RPC method for this is “**wallet_requestSilentSign**”.
* Then wallet prompts the user with a question like “XXX application wants to enable silent signing for XXX minutes/hours. Do you confirm?”
* If the user does not accept this or the RPC method is not allowed, then the application should fall back to the regular methods.
* If the user accepts, then every transaction that happens after that for the specified period will be sent by using the “**wallet_silentSendTransaction**” method.

### Implementation

There will be no change on the WalletConnect Bridge server but the client libraries and wallet should implement new RPC methods.

### New RPC Methods

#### wallet_requestSilentSign

> This RPC command will open the wallet and show a popup to the user to allow automatic signing for a given period. By calling this function the application will get a grant to call the following Commands until the timestamp expires. Standard commands like eth_signTrancaction etc will remain as now.

Reviewer 2: ```suggestion

This RPC command will open the wallet and show a popup to the user to allow automatic signing for a given period. By calling this function the application will get a grant to call the following Commands until the timestamp expires. Standard commands like `eth_signTrancaction` will remain unchanged.
```

> ```

Reviewer 3: Small typo mistake.
eth_signTrancaction ==> eth_signTransaction

```
Parameters
  clientSecret: DATA, 20 Bytes - secret key obtained from game_requestAutoSign method
  Object - The transaction object
    from: DATA, 20 Bytes - The address the transaction is sent from.
    to: DATA, 20 Bytes - (optional when creating new contract) The address the transaction is directed to.
    gas: QUANTITY - (optional, default: 90000) Integer of the gas provided for the transaction execution. It will return unused gas.
    gasPrice: QUANTITY - (optional, default: To-Be-Determined) Integer of the gasPrice used for each paid gas, in Wei.
    value: QUANTITY - (optional) Integer of the value sent with this transaction, in Wei.
    data: DATA - The compiled code of a contract OR the hash of the invoked method signature and encoded parameters.
    nonce: QUANTITY - (optional) Integer of a nonce. This allows to overwrite your own pending transactions that use the same nonce.
  
Returns
  DATA, The signed transaction object.
```

#### `wallet_silentSendTransaction`

> This command will create a transaction without asking the user and send it to the blockchain.

Reviewer 2: ```suggestion

This command will create a transaction without asking the user and send it to the blockchain.
```

> ```

Reviewer 2: ```suggestion
```shell
```

Parameters
  clientSecret: DATA, 20 Bytes - secret key obtained from game_requestAutoSign method
  Object - The transaction object
    from: DATA, 20 Bytes - The address the transaction is sent from.
    to: DATA, 20 Bytes - (optional when creating new contract) The address the transaction is directed to.
    gas: QUANTITY - (optional, default: 90000) Integer of the gas provided for the transaction execution. It will return unused gas.
    gasPrice: QUANTITY - (optional, default: To-Be-Determined) Integer of the gasPrice used for each paid gas.
    value: QUANTITY - (optional) Integer of the value sent with this transaction.
    data: DATA - The compiled code of a contract OR the hash of the invoked method signature and encoded parameters.
    nonce: QUANTITY - (optional) Integer of a nonce. This allows to overwrite your own pending transactions that use the same nonce.

Returns
  DATA, 32 Bytes - the transaction hash, or the zero hash if the transaction is not yet available.
```

### Application and Wallet Communication

> When the wallet application is closed or in the background, it cannot connect to the bridge server via WebSocket. So that we have to trigger the wallet to connect to the bridge and get waiting for requests. For this reason, the Push Notification method should be used. So only the wallets that have push notification support can implement this feature.

Reviewer 2: ```suggestion

When the wallet application is closed or in the background, it cannot connect to the bridge server via WebSocket. So that we have to trigger the wallet to connect to the bridge and get waiting for requests. For this reason, the Push Notification method should be used. So only the wallets that have push notification support can implement this feature.
```

![](../assets/eip-5345/walletconnect-flow.png)

Whenever the wallet receives a push message. It connects to the bridge server and gets the pending requests. If there is a silent signing request like “**wallet_silenSignTransaction**” or “**wallet_silentSendTransaction**” and the requesting client is allowed at that time, then the wallet will execute the request without any UI message or prompt.

## Rationale

> If we think about games and some metaverse applications, there are a lot of user interactions that need wallet interaction. In these cases, the user switches to the wallet and approves the transaction. If the user does that a lot, this will create a bad user experience.

Reviewer 2: No first person

> For this kind of applications, wallets can support a silent signing feature to allow wallets to sign transactions without user interaction.

Reviewer 2: ```suggestion
For this kind of application, wallets can support a silent signing feature to allow wallets to sign transactions without user interaction.
```

---
eip: 5345

> title: Walletconnect Silent-Signing Extension
> description: Capability of transaction signing without user interaction for a period of time for walletconnect

Reviewer 4: ```suggestion
title: WalletConnect Silent-Signing Extension
description: Capability of transaction signing without user interaction for a period of time for WalletConnect
```

> title: Walletconnect Silent-Signing Extension
> description: Capability of transaction signing without user interaction for a period of time for walletconnect

Reviewer 2: ```suggestion
title: Silent Signing Extension
description: Temporary transaction signing without user interaction
```

author: Stanley Wu (@fruit37), Mücahit Büyükyılmaz (@anndro), Muhammed Emin Aydın (@muhammedea)
discussions-to: https://ethereum-magicians.org/t/walletconnect-silent-signing-extension/10137
status: Draft
type: Standards Track

> category: Interface

Reviewer 5: Note this is already in the `interface` track, and shoudl probably be adjusted accordingly to describe functionality for ALL wallets
